classes/statistics/MetricsDAO.inc.php

```diff
@@ -94,9 +94,241 @@ protected function getAssocObjectInfo($submissionId, $contextId) {
+						ORDER BY
+							ed.stage_id DESC, ed.round DESC, ed.date_decided DESC
+						LIMIT 1
+					)
+				WHERE
+					-- skip unfinished submissions
+					s.submission_progress = 0
+					' . ($where ? ' AND ' . implode(' AND ', $where) : '') . '
+			) AS statistics
+			INNER JOIN (
+				SELECT
+					YEAR(COALESCE(?, CURRENT_TIMESTAMP))
+					- YEAR(
+						COALESCE(
+							?,
+							(
+								SELECT s.date_submitted
+								FROM submissions s
+								WHERE s.date_submitted IS NOT NULL
+								ORDER BY s.date_submitted
+								LIMIT 1
+							),
+							CURRENT_TIMESTAMP
+						)
+					) + 1 AS count
+			) AS years
+				ON 1 = 1
+		';
+		return $this->retrieve($sql, $params)->GetRowAssoc();
+	}
+
+	/**
+	 * Retrieves statistics about user registrations, with sums and averages
+	 * @param $contextId int The context id
+	 * @param $dateStart string The start date filter
+	 * @param $dateEnd string The end date filter
+	 * @return array Returns a dictionary with statistics
+	 */
+	public function getUserStatistics($contextId = null, $dateStart = null, $dateEnd = null) {
+		$where = $params = [];
+
+		if ($contextId) {
+			$where[] = 'ug.context_id = ?';
+			$params[] = $contextId;
+		}
+
+		if ($dateStart || $dateEnd) {
+			$where[] = 'u.date_registered ' . ($dateStart && $dateEnd ? 'BETWEEN ? AND ?' : ($dateStart ? '>= ?' : '<= ?'));
+			$dateStart && $params[] = $dateStart;
+			$dateEnd && $params[] = $dateEnd;
+		}
+
+		$params = array_merge($params, $params, [$dateEnd, $dateStart]);
+
+		$sql = '
+			SELECT
+				statistics.*, statistics.total / years.count AS average
+			FROM (
+				SELECT roles.role_id, COUNT(0) AS total
+				FROM (
+					-- grouping by user_id and role_id to remove duplicates
+					SELECT ug.role_id
+					FROM user_user_groups uug
+					INNER JOIN users u
+						ON u.user_id = uug.user_id
+					INNER JOIN user_groups ug
+						ON ug.user_group_id = uug.user_group_id
+					WHERE 1 = 1
+					' . ($where ? ' AND ' . implode(' AND ', $where) : '') . '
+
+					GROUP BY uug.user_id, ug.role_id
+				) roles
+				GROUP BY roles.role_id
+
+				UNION ALL
+
+				SELECT 0, COUNT(0)
+				FROM users u
+				WHERE EXISTS(
+					SELECT 0
+					FROM user_user_groups uug
+					LEFT JOIN user_groups ug
+						ON ug.user_group_id = uug.user_group_id
+					WHERE uug.user_id = u.user_id
+					' . ($where ? ' AND ' . implode(' AND ', $where) : '') . '
+				)
+			) AS statistics
+			INNER JOIN (
+				SELECT
+					YEAR(COALESCE(?, CURRENT_TIMESTAMP))
+					- YEAR(
+						COALESCE(
+							?,
+							(
+								SELECT u.date_registered
+								FROM users u
+								WHERE u.date_registered IS NOT NULL
+								ORDER BY u.date_registered
+								LIMIT 1
+							),
+							CURRENT_TIMESTAMP
+						)
+					) + 1 AS count
+			) AS years
+				ON 1 = 1
+			GROUP BY
+				role_id
+		';
+		return array_reduce($this->retrieve($sql, $params)->GetAll(), function ($data, $row) {
+			$data[$row['role_id']] = $row;
+			return $data;
+		}, []);
+	}
+}
```

Comment:
Phew! SQL is not my strong suit and this looks like some challenging work. Nice job!

When it comes time to port this work to `master`, lets have a conversation about migrating these queries to a `QueryBuilder` class. Our main goals will be to make it easier to adapt the queries for OJS/OMP, and also to create a more reusable query structure, so that plugins (or us in different use cases) can use the QueryBuilder to pull out just the information they want.

Reply:
I'll take a look in the Illuminate later to see what it offers.